File: apps/omninova-ios/OmniNovaPhoneAgent/Services/SpamDetector.swift
import Foundation

/// 骚扰电话识别器，规则来源于 `skills/phone-call-assistant/spam_detection_rules.json`。
///
/// 可在 App 启动时预置内置规则，并从网关通过 `AgentGatewayClient.fetchSpamRules()` 动态覆盖。
@MainActor @Observable
final class SpamDetector {
    enum Severity: String, Codable {
        case block, silence, allow
    }

    struct Decision: Codable, Equatable {
        let severity: Severity
        let matchedRuleId: String?
        let reason: String?

        static let allow = Decision(severity: .allow, matchedRuleId: nil, reason: nil)
    }

    struct Rule: Codable {
        let id: String
        let severity: Severity
        let reason: String
        let match: Match
    }

    struct Match: Codable {
        let type: String
        let values: [String]?
        let windowSeconds: Int?
        let threshold: Int?
        let scope: String?
        let startLocal: String?
        let endLocal: String?
        let appliesTo: String?

        enum CodingKeys: String, CodingKey {
            case type, values, threshold, scope
            case windowSeconds = "window_seconds"
            case startLocal = "start_local"
            case endLocal = "end_local"
            case appliesTo = "applies_to"
        }
    }

    struct RulesFile: Codable {
        let schemaVersion: String?
        let rules: [Rule]
        let numberBlocklist: [String]?
        let numberAllowlist: [String]?

        enum CodingKeys: String, CodingKey {
            case schemaVersion = "schema_version"
            case rules
            case numberBlocklist = "number_blocklist"
            case numberAllowlist = "number_allowlist"
        }
    }

    private(set) var rulesFile: RulesFile = RulesFile(
        schemaVersion: "1.0",
        rules: [],
        numberBlocklist: [],
        numberAllowlist: []
    )

    private var recentCalls: [String: [Date]] = [:]

    nonisolated init(bundledJSON: Data? = nil) {
        if let bundledJSON, let rf = try? JSONDecoder().decode(RulesFile.self, from: bundledJSON) {
            self.rulesFile = rf
        } else if let data = Self.loadBundledDefault() {
            if let rf = try? JSONDecoder().decode(RulesFile.self, from: data) {
                self.rulesFile = rf
            }
        }
    }

    func updateRules(from data: Data) {
        guard let rf = try? JSONDecoder().decode(RulesFile.self, from: data) else { return }
        self.rulesFile = rf
    }

    /// 评估一通来电。`firstUtterance` 可在前几秒 ASR 拿到开场白后补充调用。
    func evaluate(
        phoneNumber: String?,
        isInContacts: Bool,
        firstUtterance: String?,
        now: Date = Date()
    ) -> Decision {
        if let phone = phoneNumber, rulesFile.numberAllowlist?.contains(phone) == true {
            return .allow
        }
        if let phone = phoneNumber, rulesFile.numberBlocklist?.contains(phone) == true {
            return Decision(severity: .block, matchedRuleId: "blocklist", reason: "硬黑名单")
        }

        if let phone = phoneNumber {
            recentCalls[phone, default: []].append(now)
            recentCalls[phone] = recentCalls[phone]?.filter { now.timeIntervalSince($0) < 24 * 3600 }
        }

        for rule in rulesFile.rules {
            if matches(rule: rule, phone: phoneNumber, isInContacts: isInContacts,
                       utterance: firstUtterance, now: now) {
                return Decision(severity: rule.severity, matchedRuleId: rule.id, reason: rule.reason)
            }
        }
        return .allow
    }

    private func matches(
        rule: Rule,
        phone: String?,
        isInContacts: Bool,
        utterance: String?,
        now: Date
    ) -> Bool {
        switch rule.match.type {
        case "prefix":
            guard let phone, let values = rule.match.values else { return false }
            return values.contains { phone.hasPrefix($0) }
        case "keyword":
            guard let utterance = utterance?.lowercased(), let values = rule.match.values else {
                return false
            }
            return values.contains { utterance.contains($0.lowercased()) }
        case "rate_limit":
            guard let phone,
                  let window = rule.match.windowSeconds,
                  let threshold = rule.match.threshold else { return false }
            if rule.match.scope == "not_in_contacts" && isInContacts { return false }
            let history = recentCalls[phone] ?? []
            let recent = history.filter { now.timeIntervalSince($0) < TimeInterval(window) }
            return recent.count >= threshold
        case "time_window":
            guard let start = rule.match.startLocal,
                  let end = rule.match.endLocal else { return false }
            if rule.match.scope == "not_in_contacts" && isInContacts { return false }
            let cal = Calendar.current
            let comps = cal.dateComponents([.hour, .minute], from: now)
            let minutes = (comps.hour ?? 0) * 60 + (comps.minute ?? 0)
            let startMin = Self.parseMinutes(start)
            let endMin = Self.parseMinutes(end)
            if startMin > endMin {
                return minutes >= startMin || minutes <= endMin
            } else {
                return minutes >= startMin && minutes <= endMin
            }
        default:
            return false
        }
    }

    private static func parseMinutes(_ hhmm: String) -> Int {
        let parts = hhmm.split(separator: ":")
        guard parts.count == 2, let h = Int(parts[0]), let m = Int(parts[1]) else { return 0 }
        return h * 60 + m
    }

    nonisolated private static func loadBundledDefault() -> Data? {
        guard let url = Bundle.main.url(forResource: "spam_detection_rules", withExtension: "json") else {
            return nil
        }
        return try? Data(contentsOf: url)
    }
}

File: novalclaw-main/apps/omninova-android/app/src/main/java/com/omninova/phoneagent/call/SpamDetector.kt
package com.omninova.phoneagent.call

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import java.time.LocalTime

/**
 * 骚扰识别，规则与 `skills/phone-call-assistant/spam_detection_rules.json` 对齐。
 */
class SpamDetector(initialJson: String? = null) {

    @Serializable
    enum class Severity { block, silence, allow }

    @Serializable
    data class Match(
        val type: String,
        val values: List<String>? = null,
        @SerialName("window_seconds") val windowSeconds: Int? = null,
        val threshold: Int? = null,
        val scope: String? = null,
        @SerialName("start_local") val startLocal: String? = null,
        @SerialName("end_local") val endLocal: String? = null,
        @SerialName("applies_to") val appliesTo: String? = null,
    )

    @Serializable
    data class Rule(
        val id: String,
        val severity: Severity,
        val reason: String,
        val match: Match,
    )

    @Serializable
    data class RulesFile(
        @SerialName("schema_version") val schemaVersion: String = "1.0",
        val rules: List<Rule> = emptyList(),
        @SerialName("number_blocklist") val numberBlocklist: List<String> = emptyList(),
        @SerialName("number_allowlist") val numberAllowlist: List<String> = emptyList(),
    )

    data class Decision(
        val severity: Severity,
        val matchedRuleId: String? = null,
        val reason: String? = null,
    ) {
        companion object {
            val ALLOW = Decision(Severity.allow)
        }
    }

    private val parser = Json { ignoreUnknownKeys = true }

    @Volatile
    var rulesFile: RulesFile = initialJson?.let {
        runCatching { parser.decodeFromString(RulesFile.serializer(), it) }.getOrNull()
    } ?: RulesFile()
        private set

    private val recent = HashMap<String, ArrayDeque<Long>>()

    fun updateRules(jsonText: String) {
        runCatching { parser.decodeFromString(RulesFile.serializer(), jsonText) }
            .onSuccess { rulesFile = it }
    }

    fun evaluate(
        phoneNumber: String?,
        isInContacts: Boolean,
        firstUtterance: String?,
        nowMs: Long = System.currentTimeMillis(),
        now: LocalTime = LocalTime.now(),
    ): Decision {
        if (phoneNumber != null) {
            if (rulesFile.numberAllowlist.contains(phoneNumber)) return Decision.ALLOW
            if (rulesFile.numberBlocklist.contains(phoneNumber)) {
                return Decision(Severity.block, "blocklist", "硬黑名单")
            }
            val history = recent.getOrPut(phoneNumber) { ArrayDeque() }
            history.addLast(nowMs)
            while (history.isNotEmpty() && nowMs - history.first() > 86_400_000L) {
                history.removeFirst()
            }
        }

        for (rule in rulesFile.rules) {
            if (matches(rule, phoneNumber, isInContacts, firstUtterance, nowMs, now)) {
                return Decision(rule.severity, rule.id, rule.reason)
            }
        }
        return Decision.ALLOW
    }

    private fun matches(
        rule: Rule,
        phone: String?,
        isInContacts: Boolean,
        utterance: String?,
        nowMs: Long,
        now: LocalTime,
    ): Boolean {
        return when (rule.match.type) {
            "prefix" -> {
                val p = phone ?: return false
                rule.match.values?.any { p.startsWith(it) } ?: false
            }
            "keyword" -> {
                val u = utterance?.lowercase() ?: return false
                rule.match.values?.any { u.contains(it.lowercase()) } ?: false
            }
            "rate_limit" -> {
                val p = phone ?: return false
                val window = rule.match.windowSeconds ?: return false
                val threshold = rule.match.threshold ?: return false
                if (rule.match.scope == "not_in_contacts" && isInContacts) return false
                val history = recent[p] ?: return false
                val cutoff = nowMs - window * 1000L
                history.count { it >= cutoff } >= threshold
            }
            "time_window" -> {
                val start = rule.match.startLocal ?: return false
                val end = rule.match.endLocal ?: return false
                if (rule.match.scope == "not_in_contacts" && isInContacts) return false
                val s = LocalTime.parse(start)
                val e = LocalTime.parse(end)
                if (s.isAfter(e)) !now.isBefore(s) || !now.isAfter(e)
                else !now.isBefore(s) && !now.isAfter(e)
            }
            else -> false
        }
    }
}
